File: src/main/java/eu/dragoncoding/dragonbot/structures/CriteriaPackage.java
package eu.dragoncoding.dragonbot.structures;

import javax.persistence.criteria.CriteriaBuilder;
import javax.persistence.criteria.CriteriaQuery;
import javax.persistence.criteria.Root;

public class CriteriaPackage<T extends Object & Entity> {
    final CriteriaBuilder criteriaBuilder;
    final CriteriaQuery<T> query;
    final Root<T> root;

    public CriteriaPackage(CriteriaBuilder builder, Class<T> clazz) {
        criteriaBuilder = builder;
        query = builder.createQuery(clazz);
        root = query.from(clazz);
    }

    public CriteriaBuilder getCriteriaBuilder() {
        return criteriaBuilder;
    }
    public CriteriaQuery<T> getQuery() {
        return query;
    }
    public Root<T> getRoot() {
        return root;
    }
}

File: src/main/java/eu/dragoncoding/dragonbot/structures/CriteriaPackage.kt
package eu.dragoncoding.dragonbot.structures

import javax.persistence.criteria.CriteriaBuilder
import javax.persistence.criteria.CriteriaQuery
import javax.persistence.criteria.Root

class CriteriaPackage<T>(val criteriaBuilder: CriteriaBuilder, clazz: Class<T>) where T : Any, T : Entity {
    val query: CriteriaQuery<T> = criteriaBuilder.createQuery(clazz)
    val root: Root<T> = query.from(clazz)
}